### ISSUE TITLE
How to build kpatch module when CONFIG_LIVEPATCH is enabled?

### ISSUE THREAD

**Issue by aaronlu** [dynup/kpatch#947]
For reasons like kpatch_shadow_* and existing deployed kpatch modules, I would like to build kpatch modules even for kernels with CONFIG_LIVEPATCH=y.

I don't see how to do this after reading README and taking a quick look at Makefile, greping LIVEPATCH shows that KPATCH_MODULE seems to be deciding which format to build. Is it that I should modify this value in this file to get what I want, or there is an easier way to force a kpatch module build? Thanks.

**Comment by aaronlu** [comment]
Thanks for the info.

BTW, I tried to do a livepatch build with kpatch_shadow_* used in patch, I got this:
... ...
Reading special section data
Building original source
Building patched source
Extracting new and modified ELF sections
page_io.o: changed function: generic_swapfile_activate
swapfile.o: new function: offset_in_range
swapfile.o: changed function: map_swap_entry
swapfile.o: changed function: destroy_swap_extents
swapfile.o: changed function: swap_do_scheduled_discard
Patched objects: vmlinux
Building patch module: livepatch-swap_extents_shadow.ko
ERROR: Undefined symbols: kpatch_shadow_alloc kpatch_shadow_free kpatch_shadow_get . Check /root/.kpatch/build.log for more details.

Is this problem known? Or perhaps I use kpatch_shadow_* the wrong way?

I got to know shadow API here:
https://github.com/dynup/kpatch/blob/master/doc/patch-author-guide.md#use-a-shadow-variable
And right above, it said "The guidance in this document applies only to kpatch.":
https://github.com/dynup/kpatch/blob/master/doc/patch-author-guide.md#kpatch-vs-livepatch-vs-kgraft
So I thought shadow is only supported in kpatch module. Maybe I missed something?

**Comment by joe-lawrence** [comment]
Unfortunately the upstream shadow variable API is different its the kpatch cousin. 

See [Documentation/livepatch/shadow-vars.txt](https://github.com/torvalds/linux/blob/master/Documentation/livepatch/shadow-vars.txt) for a summary of the kernel's shadow variable functions.

**Comment by aaronlu** [comment]
OK I see, and it's only available from v4.15...
Looks like I may need to fight with kpatch module on our 4.9 kernel

**Comment by jpoimboe** [comment]
Yeah, if you need shadow variables for 4.9 then I guess you'll probably still need kpatch.ko.

**Comment by aaronlu** [comment]
With some tweaks to the current head of the tree, I can build kpatch.ko and kpatch-my.ko even with CONFIG_LIVEPATCH=y.

But the built kpatch-my.ko can not work with the existing kpatch.ko on the server. The exising kpatch.ko is long loaded before, so should be built with a much older tree I guess. And there are some other kpatch-XXX.ko depending on it, so unload it is not an option.

In the meantime, the newly built kpatch-my.ko can only work with the newly built kpatch.ko: with the old kpatch.ko, when I tried to load kpatch-my.ko, I would get:
[10911332.850703] kpatch_my: module verification failed: signature and/or required key missing - tainting kernel
[10911332.853038] kpatch_my: disagrees about version of symbol kpatch_register
[10911332.854171] kpatch_my: Unknown symbol kpatch_register (err -22)
[10911332.855285] kpatch_my: disagrees about version of symbol kpatch_unregister
[10911332.856396] kpatch_my: Unknown symbol kpatch_unregister (err -22)
[10911332.857493] kpatch_my: Unknown symbol kpatch_root_kobj (err 0)

Guess I'm doomed.

**Comment by jpoimboe** [comment]
When patching a previously patched system, in general it's a good idea to keep kpatch-build and kpatch.ko at the same git tree level as the original patch.  I guess you don't know what level of the code was used before?

If not, I guess you could try reverting to older versions of kmod/core/core.c until it works.

**Comment by aaronlu** [comment]
I should have made it more clear, sorry for the confusion.

I have the old tree, but it failed to build the module due to create-diff-object complains:
ERROR: swapfile.o: symbol changed sections: .LC1

I figured out this is solved by you:
commit b1095f485ec41f083bb266ad0810cc6fa8fc1253
Author: Josh Poimboeuf <jpoimboe@redhat.com>
Date:   Mon Aug 21 09:48:03 2017 -0500

    create-diff-object: handle .LCx symbol renames

so I backported to the old tree but then there is new problem:
swapfile.o: changed section .rela.discard.retpoline_safe not selected for inclusion
/usr/local/libexec/kpatch/create-diff-object: unreconcilable difference

I haven't been able to figure out which commit fixed this.

I also thought about using the latest create-diff-object with the old tree, but the latest create-diff-object requires more params now and I haven't figured out how to adjust these things yet and whether using latest create-diff-object with old tree will produce correct result.

**Comment by jpoimboe** [comment]
Try also backporting commit 9b2e37ba.

**Comment by aaronlu** [comment]
That's definitely one step closer, thanks!
Now the error became:
/usr/local/libexec/kpatch/create-diff-object: ERROR: swapfile.o: kpatch_create_dynamic_rela_sections: 2444: lookup_local_symbol .LC1 (swapfile.c) needed for __bug_table

**Comment by jpoimboe** [comment]
Hm, that one is not obvious to me.  Another crazy idea is to just use livepatch in addition to kpatch.ko.  I can't think of any reason why livepatch and kpatch can't coexist, though I've never tried it :-)

**Comment by aaronlu** [comment]
Thanks a lot for your kind help!
I have actually tried it, the livepatch module can be loaded and work as expected on that server, I don't see any immediate problem :-)
